src/model.py

```diff
@@ -175,30 +215,130 @@ def build_bow(self, corpus):
+            b = BookText(rawtext = corpus_vals[i])
             bb = b.clean(deromanize=True, lemmatize=True) 
             words += bb.tokenize(on='words', rem_stopwords=True)
+        
         words = [w.lower() for w in words]
+        
+       
+        
+        
+        #---------------------------remove too frequent and too infrequent words-----------------
+        
+        dist = FreqDist(words)
+        repeats = {}
+        for rep_count in sorted(list(set(dist.values())), reverse = True):
+            if (rep_count >=self.min_occurences) & (rep_count <=self.max_occurences):
+                repeats[rep_count] = [word for word in dist.keys() if dist[word] == rep_count]
+                
+        words_minus_tails = []
+        for v in repeats.values():
+            words_minus_tails += v 
+        words = words_minus_tails
+        
+       
+        
 
-        words = list(set(words)) #Making a set early on reduces the size to speed things up
-
-
-        #---------------STEP-2: remove all numeric characters ------------------------------------
+        #----------------------------remove all numeric characters ------------------------------
         no_nums = []
         for word in words:
             if (not any(ch.isdigit() for ch in word)):
                 no_nums.append(word)
-
-        #---------------STEP-3: Package as a dictionary ------------------------------------------
-        bow = no_nums
+                
+        all_words = no_nums
+        
+        
+        
+        
+        #--------------------------- identify and remove all proper nouns ------------------------
+        
+        if self.include_proper_nouns:
+            proper_nouns = []
+        
+        else:
+            proper_nouns= self.proper_nouns_in_corpus(corpus) 
+            proper_nouns = [word.lower() for word in proper_nouns]
+        
+        words_minus_proper = [word for word in all_words if word not in proper_nouns]
+        
+        #------------------------Package as a dictionary ------------------------------------------
         vocab_dict = {}
-        for i in range(len(bow)):
-            vocab_dict[bow[i]] = i
+        for i in range(len(words_minus_proper)):
+            vocab_dict[words_minus_proper[i]] = i
         
         self.bow = vocab_dict
+        
+        #------------------------------fin.------------------------------------------------------
+
+        
+        
+        
+    def proper_nouns_in_corpus(self, corpus: pd.Series):
+        
+        """ collects all proper nouns in the corpus and returns them as a list"""
+
+        def find_proper_nouns(text):
+            """ A local helper function """
+            bt = BookText(rawtext=text)
+            bt.clean(lemmatize=True, deromanize=True,inplace=True)
+            sentences = bt.tokenize(on = 'sent') #all sentences in one row
+            tagged_sentences = [pos_tag(sentence.split()) for sentence in sentences]  # tagged sentences
```

Comment:
I'm not sure it matters, but `nltk.pos_tag_sents` is allegedly a lot faster than `pos_tag`

Reply:
I looked into it and tried incorporating it into the code. 
Unfortunately because of the way bt.tokenize returns the sentences it cannot be easily implemented without adding an additional data massaging step. This step completely defeats the tiny gains in speed from switching to pos_tag_sents.

I  am choosing to keep it as is.